File: 24_高级数据结构/good python/ds/lazy_avl_tree.py
# from titan_pylib.data_structures.avl_tree.lazy_avl_tree import LazyAVLTree
from typing import Generic, Iterable, TypeVar, Callable, List, Tuple, Optional

T = TypeVar("T")
F = TypeVar("F")


class LazyAVLTree(Generic[T, F]):
    """遅延伝播反転可能平衡二分木です。"""

    class Node:
        def __init__(self, key: T, id: F):
            self.key: T = key
            self.data: T = key
            self.left: Optional[LazyAVLTree.Node] = None
            self.right: Optional[LazyAVLTree.Node] = None
            self.lazy: F = id
            self.rev: int = 0
            self.height: int = 1
            self.size: int = 1

        def __str__(self):
            if self.left is None and self.right is None:
                return f"key:{self.key, self.height, self.size, self.data, self.lazy, self.rev}\n"
            return f"key:{self.key, self.height, self.size, self.data, self.lazy, self.rev},\n left:{self.left},\n right:{self.right}\n"

    def __init__(
        self,
        a: Iterable[T],
        op: Callable[[T, T], T],
        mapping: Callable[[F, T], T],
        composition: Callable[[F, F], F],
        e: T,
        id: F,
        node: Node = None,
    ) -> None:
        self.root: Optional[LazyAVLTree.Node] = node
        self.op: Callable[[T, T], T] = op
        self.mapping: Callable[[F, T], T] = mapping
        self.composition: Callable[[F, F], F] = composition
        self.e: T = e
        self.id: F = id
        a = list(a)
        if a:
            self._build(a)

    def _build(self, a: List[T]) -> None:
        Node = LazyAVLTree.Node
        id = self.id

        def sort(l: int, r: int) -> Node:
            mid = (l + r) >> 1
            node = Node(a[mid], id)
            if l != mid:
                node.left = sort(l, mid)
            if mid + 1 != r:
                node.right = sort(mid + 1, r)
            self._update(node)
            return node

        self.root = sort(0, len(a))

    def _propagate(self, node: Node) -> None:
        l, r = node.left, node.right
        if node.rev:
            node.left, node.right = r, l
            if l:
                l.rev ^= 1
            if r:
                r.rev ^= 1
            node.rev = 0
        if node.lazy != self.id:
            lazy = node.lazy
            if l:
                l.data = self.mapping(lazy, l.data)
                l.key = self.mapping(lazy, l.key)
                l.lazy = lazy if l.lazy == self.id else self.composition(lazy, l.lazy)
            if r:
                r.data = self.mapping(lazy, r.data)
                r.key = self.mapping(lazy, r.key)
                r.lazy = lazy if r.lazy == self.id else self.composition(lazy, r.lazy)
            node.lazy = self.id

    def _update(self, node: Node) -> None:
        node.size = 1
        node.data = node.key
        node.height = 1
        if node.left:
            node.size += node.left.size
            node.data = self.op(node.left.data, node.data)
            node.height = max(node.left.height + 1, 1)
        if node.right:
            node.size += node.right.size
            node.data = self.op(node.data, node.right.data)
            node.height = max(node.height, node.right.height + 1)

    def _get_balance(self, node: Node) -> int:
        return (
            (0 if node.right is None else -node.right.height)
            if node.left is None
            else (node.left.height if node.right is None else node.left.height - node.right.height)
        )

    def _balance_left(self, node: Node) -> Node:
        self._propagate(node.left)
        if node.left.left is None or node.left.left.height + 2 == node.left.height:
            u = node.left.right
            self._propagate(u)
            node.left.right = u.left
            u.left = node.left
            node.left = u.right
            u.right = node
            self._update(u.left)
        else:
            u = node.left
            node.left = u.right
            u.right = node
        self._update(u.right)
        self._update(u)
        return u

    def _balance_right(self, node: Node) -> Node:
        self._propagate(node.right)
        if node.right.right is None or node.right.right.height + 2 == node.right.height:
            u = node.right.left
            self._propagate(u)
            node.right.left = u.right
            u.right = node.right
            node.right = u.left
            u.left = node
            self._update(u.right)
        else:
            u = node.right
            node.right = u.left
            u.left = node
        self._update(u.left)
        self._update(u)
        return u

    def _kth_elm(self, k: int) -> T:
        if k < 0:
            k += len(self)
        node = self.root
        while True:
            self._propagate(node)
            t = 0 if node.left is None else node.left.size
            if t == k:
                return node.key
            elif t < k:
                k -= t + 1
                node = node.right
            else:
                node = node.left

    def _merge_with_root(self, l: Node, root: Node, r: Node) -> Node:
        diff = (
            (0 if r is None else -r.height)
            if l is None
            else (l.height if r is None else l.height - r.height)
        )
        if diff > 1:
            self._propagate(l)
            l.right = self._merge_with_root(l.right, root, r)
            self._update(l)
            if -l.right.height if l.left is None else l.left.height - l.right.height == -2:
                return self._balance_right(l)
            return l
        elif diff < -1:
            self._propagate(r)
            r.left = self._merge_with_root(l, root, r.left)
            self._update(r)
            if r.left.height if r.right is None else r.left.height - r.right.height == 2:
                return self._balance_left(r)
            return r
        else:
            root.left = l
            root.right = r
            self._update(root)
            return root

    def _merge_node(self, l: Node, r: Node) -> Node:
        if l is None:
            return r
        if r is None:
            return l
        l, tmp = self._pop_max(l)
        return self._merge_with_root(l, tmp, r)

    def merge(self, other: "LazyAVLTree") -> None:
        self.root = self._merge_node(self.root, other.node)

    def _pop_max(self, node: Node) -> Tuple[Node, Node]:
        self._propagate(node)
        path = []
        mx = node
        while node.right:
            path.append(node)
            mx = node.right
            node = node.right
            self._propagate(node)
        path.append(node.left)
        for _ in range(len(path) - 1):
            node = path.pop()
            if node is None:
                path[-1].right = None
                self._update(path[-1])
                continue
            b = self._get_balance(node)
            path[-1].right = (
                self._balance_left(node)
                if b == 2
                else self._balance_right(node)
                if b == -2
                else node
            )
            self._update(path[-1])
        if path[0]:
            b = self._get_balance(path[0])
            path[0] = (
                self._balance_left(path[0])
                if b == 2
                else self._balance_right(path[0])
                if b == -2
                else path[0]
            )
        mx.left = None
        self._update(mx)
        return path[0], mx

    def _split_node(self, node: Node, k: int) -> Tuple[Node, Node]:
        if not node:
            return None, None
        self._propagate(node)
        tmp = k if node.left is None else k - node.left.size
        if tmp == 0:
            return node.left, self._merge_with_root(None, node, node.right)
        elif tmp < 0:
            s, t = self._split_node(node.left, k)
            return s, self._merge_with_root(t, node, node.right)
        else:
            s, t = self._split_node(node.right, tmp - 1)
            return self._merge_with_root(node.left, node, s), t

    def split(self, k: int) -> Tuple["LazyAVLTree", "LazyAVLTree"]:
        l, r = self._split_node(self.root, k)
        return LazyAVLTree(
            [], self.op, self.mapping, self.composition, self.e, self.id, l
        ), LazyAVLTree([], self.op, self.mapping, self.composition, self.e, self.id, r)

    def insert(self, k: int, key: T) -> None:
        s, t = self._split_node(self.root, k)
        self.root = self._merge_with_root(s, LazyAVLTree.Node(key, self.id), t)

    def pop(self, k: int) -> T:
        s, t = self._split_node(self.root, k + 1)
        s, tmp = self._pop_max(s)
        self.root = self._merge_node(s, t)
        return tmp.key

    def apply(self, l: int, r: int, f: F) -> None:
        if l >= r or (not self.root):
            return
        stack = [(self.root), (self.root, 0, len(self))]
        while stack:
            if isinstance(stack[-1], tuple):
                node, left, right = stack.pop()
                if right <= l or r <= left:
                    continue
                self._propagate(node)
                if l <= left and right < r:
                    node.key = self.mapping(f, node.key)
                    node.data = self.mapping(f, node.data)
                    node.lazy = f if node.lazy == self.id else self.composition(f, node.lazy)
                else:
                    lsize = node.left.size if node.left else 0
                    stack.append(node)
                    if node.left:
                        stack.append((node.left, left, left + lsize))
                    if l <= left + lsize < r:
                        node.key = self.mapping(f, node.key)
                    if node.right:
                        stack.append((node.right, left + lsize + 1, right))
            else:
                self._update(stack.pop())

    def all_apply(self, f: F) -> None:
        if not self.root:
            return
        self.root.key = self.mapping(f, self.root.key)
        self.root.data = self.mapping(f, self.root.data)
        self.root.lazy = f if self.root.lazy == self.id else self.composition(f, self.root.lazy)

    def reverse(self, l: int, r: int) -> None:
        if l >= r:
            return
        s, t = self._split_node(self.root, r)
        r, s = self._split_node(s, l)
        s.rev ^= 1
        self.root = self._merge_node(self._merge_node(r, s), t)

    def all_reverse(self) -> None:
        if self.root is None:
            return
        self.root.rev ^= 1

    def prod(self, l: int, r: int) -> T:
        if l >= r or (not self.root):
            return self.e

        def dfs(node: LazyAVLTree.Node, left: int, right: int) -> T:
            if right <= l or r <= left:
                return self.e
            self._propagate(node)
            if l <= left and right < r:
                return node.data
            lsize = node.left.size if node.left else 0
            res = self.e
            if node.left:
                res = dfs(node.left, left, left + lsize)
            if l <= left + lsize < r:
                res = self.op(res, node.key)
            if node.right:
                res = self.op(res, dfs(node.right, left + lsize + 1, right))
            return res

        return dfs(self.root, 0, len(self))

    def all_prod(self) -> T:
        return self.root.data if self.root else self.e

    def clear(self) -> None:
        self.root = None

    def tolist(self) -> List[T]:
        node = self.root
        stack = []
        a = []
        while stack or node:
            if node:
                self._propagate(node)
                stack.append(node)
                node = node.left
            else:
                node = stack.pop()
                a.append(node.key)
                node = node.right
        return a

    def __len__(self):
        return 0 if self.root is None else self.root.size

    def __iter__(self):
        self.__iter = 0
        return self

    def __next__(self):
        if self.__iter == len(self):
            raise StopIteration
        res = self[self.__iter]
        self.__iter += 1
        return res

    def __reversed__(self):
        for i in range(len(self)):
            yield self[-i - 1]

    def __bool__(self):
        return self.root is not None

    def __getitem__(self, k: int) -> T:
        return self._kth_elm(k)

    def __setitem__(self, k, key: T):
        if k < 0:
            k += len(self)
        node = self.root
        path = []
        while True:
            self._propagate(node)
            path.append(node)
            t = 0 if node.left is None else node.left.size
            if t == k:
                node.key = key
                break
            if t < k:
                k -= t + 1
                node = node.right
            else:
                node = node.left
        while path:
            self._update(path.pop())

    def __str__(self):
        return "[" + ", ".join(map(str, self.tolist())) + "]"

    def __repr__(self):
        return f"LazyAVLTree({self})"

File: 24_高级数据结构/avl/lazyAVLTree.go
// api:
//   NewLazyAVLTree(n int32, f func(i int32) E) *LazyAVLTree
//   Merge(other *LazyAVLTree)
//   Split(k int32) (*LazyAVLTree, *LazyAVLTree)
//   Insert(k int32, key E)
//   Pop(k int32) E
//   Get(k int32) E
//   Set(k int32, key E)
//   Update(start, end int32, f Id)
//   UpdateAll(f Id)
//   Query(start, end int32) E
//   QueryAll() E
//   Clear()
//   ToList() []E
//   Size() int32

package main

import (
	"fmt"
	"math/rand"
	"runtime/debug"
	"time"
)

func main() {
	test()
	testTime()
}

func init() {
	debug.SetGCPercent(-1)
}

type E = struct {
	sum  int
	size int32
}
type Id = int

func e() E   { return E{} }
func id() Id { return 0 }
func op(x, y E) E {
	x.sum += y.sum
	x.size += y.size
	return x
}
func mapping(f Id, x E) E {
	x.sum += int(x.size) * f
	return x
}
func composition(f, g Id) Id { return f + g }

type avlNode struct {
	key         E
	data        E
	left, right *avlNode
	lazy        Id
	height      int8
	size        int32
}

func _newAVLNode(key E, id Id) *avlNode {
	return &avlNode{key: key, data: key, lazy: id, height: 1, size: 1}
}

var (
	tmpPath  = make([]*avlNode, 0, 128)
	tmpStack = make([]stackItem, 0, 128)
)

type stackItem struct {
	node        *avlNode
	left, right int32
}
type LazyAVLTree struct {
	root *avlNode
}

func NewLazyAVLTree(n int32, f func(i int32) E) *LazyAVLTree {
	res := &LazyAVLTree{}
	if n > 0 {
		res._build(n, f)
	}
	return res
}

func _newWithRoot(root *avlNode) *LazyAVLTree {
	return &LazyAVLTree{root: root}
}

func (avl *LazyAVLTree) Merge(other *LazyAVLTree) {
	avl.root = avl._mergeNode(avl.root, other.root)
}

func (avl *LazyAVLTree) Split(k int32) (*LazyAVLTree, *LazyAVLTree) {
	l, r := avl._splitNode(avl.root, k)
	return _newWithRoot(l), _newWithRoot(r)
}

func (avl *LazyAVLTree) Insert(k int32, key E) {
	if k < 0 {
		k += avl.Size()
	}
	if k < 0 {
		k = 0
	}
	if n := avl.Size(); k > n {
		k = n
	}
	s, t := avl._splitNode(avl.root, k)
	avl.root = avl._mergeWithRoot(s, _newAVLNode(key, id()), t)
}

func (avl *LazyAVLTree) Pop(k int32) E {
	if k < 0 {
		k += avl.Size()
	}
	s, t := avl._splitNode(avl.root, k+1)
	s, tmp := avl._popMax(s)
	avl.root = avl._mergeNode(s, t)
	return tmp.key
}

func (avl *LazyAVLTree) Get(k int32) E {
	return avl._kthElm(k)
}

func (avl *LazyAVLTree) Set(k int32, key E) {
	if k < 0 {
		k += avl.Size()
	}
	node := avl.root
	tmpPath = tmpPath[:0]
	path := tmpPath
	for {
		avl._propagate(node)
		path = append(path, node)
		t := int32(0)
		if node.left != nil {
			t = node.left.size
		}
		if t == k {
			node.key = key
			break
		}
		if t < k {
			k -= t + 1
			node = node.right
		} else {
			node = node.left
		}
	}
	for i := len(path) - 1; i >= 0; i-- {
		avl._update(path[i])
	}
}

func (avl *LazyAVLTree) Update(start, end int32, f Id) {
	if start < 0 {
		start = 0
	}
	if n := avl.Size(); end > n {
		end = n
	}
	if start >= end || avl.root == nil {
		return
	}

	tmpStack = tmpStack[:0]
	stack := tmpStack
	root := avl.root
	stack = append(stack, stackItem{root, -1, -1}, stackItem{root, 0, avl.Size()})
	for len(stack) > 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		node, left, right := top.node, top.left, top.right
		if left == -1 {
			avl._update(node)
			continue
		}
		if right <= start || end <= left {
			continue
		}
		avl._propagate(node)
		if start <= left && right < end {
			node.key = mapping(f, node.key)
			node.data = mapping(f, node.data)
			if node.lazy == id() {
				node.lazy = f
			} else {
				node.lazy = composition(f, node.lazy)
			}
		} else {
			lsize := int32(0)
			if node.left != nil {
				lsize = node.left.size
			}
			stack = append(stack, stackItem{node, -1, -1})
			if node.left != nil {
				stack = append(stack, stackItem{node.left, left, left + lsize})
			}
			if tmp := left + lsize; start <= tmp && tmp < end {
				node.key = mapping(f, node.key)
			}
			if node.right != nil {
				stack = append(stack, stackItem{node.right, left + lsize + 1, right})
			}
		}
	}
}

func (avl *LazyAVLTree) UpdateAll(f Id) {
	if avl.root == nil {
		return
	}
	root := avl.root
	root.key = mapping(f, root.key)
	root.data = mapping(f, root.data)
	if root.lazy == id() {
		root.lazy = f
	} else {
		root.lazy = composition(f, root.lazy)
	}
}

func (avl *LazyAVLTree) Query(start, end int32) E {
	if start < 0 {
		start = 0
	}
	if n := avl.Size(); end > n {
		end = n
	}
	if start >= end || avl.root == nil {
		return e()
	}
	var dfs func(node *avlNode, left, right int32) E
	dfs = func(node *avlNode, left, right int32) E {
		if right <= start || end <= left {
			return e()
		}
		avl._propagate(node)
		if start <= left && right < end {
			return node.data
		}
		lsize := int32(0)
		if node.left != nil {
			lsize = node.left.size
		}
		res := e()
		if node.left != nil {
			res = dfs(node.left, left, left+lsize)
		}
		if tmp := left + lsize; start <= tmp && tmp < end {
			res = op(res, node.key)
		}
		if node.right != nil {
			res = op(res, dfs(node.right, left+lsize+1, right))
		}
		return res
	}
	return dfs(avl.root, 0, avl.Size())
}

func (avl *LazyAVLTree) QueryAll() E {
	if avl.root == nil {
		return e()
	}
	return avl.root.data
}

func (avl *LazyAVLTree) Clear() { avl.root = nil }

func (avl *LazyAVLTree) ToList() []E {
	node := avl.root
	stack := []*avlNode{}
	res := make([]E, 0, avl.Size())
	for len(stack) > 0 || node != nil {
		if node != nil {
			avl._propagate(node)
			stack = append(stack, node)
			node = node.left
		} else {
			node = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res = append(res, node.key)
			node = node.right
		}
	}
	return res
}

func (avl *LazyAVLTree) Size() int32 {
	if avl.root == nil {
		return 0
	}
	return avl.root.size
}

func (avl *LazyAVLTree) _build(n int32, f func(i int32) E) {
	var dfs func(l, r int32) *avlNode
	dfs = func(l, r int32) *avlNode {
		mid := (l + r) >> 1
		node := _newAVLNode(f(mid), id())
		if l != mid {
			node.left = dfs(l, mid)
		}
		if mid+1 != r {
			node.right = dfs(mid+1, r)
		}
		avl._update(node)
		return node
	}
	avl.root = dfs(0, n)
}

func (avl *LazyAVLTree) _propagate(node *avlNode) {
	l, r := node.left, node.right
	if node.lazy != id() {
		lazy := node.lazy
		if l != nil {
			l.data = mapping(lazy, l.data)
			l.key = mapping(lazy, l.key)
			if l.lazy == id() {
				l.lazy = lazy
			} else {
				l.lazy = composition(lazy, l.lazy)
			}
		}
		if r != nil {
			r.data = mapping(lazy, r.data)
			r.key = mapping(lazy, r.key)
			if r.lazy == id() {
				r.lazy = lazy
			} else {
				r.lazy = composition(lazy, r.lazy)
			}
		}
		node.lazy = id()
	}
}

func (avl *LazyAVLTree) _update(node *avlNode) {
	node.size = 1
	node.data = node.key
	node.height = 1
	if node.left != nil {
		node.size += node.left.size
		node.data = op(node.left.data, node.data)
		node.height = max8(node.left.height+1, 1)
	}
	if node.right != nil {
		node.size += node.right.size
		node.data = op(node.data, node.right.data)
		node.height = max8(node.height, node.right.height+1)
	}
}

func (avl *LazyAVLTree) _getBalance(node *avlNode) int8 {
	if node.left == nil {
		if node.right == nil {
			return 0
		}
		return -node.right.height
	}
	if node.right == nil {
		return node.left.height
	}
	return node.left.height - node.right.height
}

func (avl *LazyAVLTree) _balanceLeft(node *avlNode) *avlNode {
	avl._propagate(node.left)
	var u *avlNode
	if node.left.left == nil || node.left.left.height+2 == node.left.height {
		u = node.left.right
		avl._propagate(u)
		node.left.right = u.left
		u.left = node.left
		node.left = u.right
		u.right = node
		avl._update(u.left)
	} else {
		u = node.left
		node.left = u.right
		u.right = node
	}
	avl._update(u.right)
	avl._update(u)
	return u
}

func (avl *LazyAVLTree) _balanceRight(node *avlNode) *avlNode {
	avl._propagate(node.right)
	var u *avlNode
	if node.right.right == nil || node.right.right.height+2 == node.right.height {
		u = node.right.left
		avl._propagate(u)
		node.right.left = u.right
		u.right = node.right
		node.right = u.left
		u.left = node
		avl._update(u.right)
	} else {
		u = node.right
		node.right = u.left
		u.left = node
	}
	avl._update(u.left)
	avl._update(u)
	return u
}

func (avl *LazyAVLTree) _kthElm(k int32) E {
	if k < 0 {
		k += avl.Size()
	}
	node := avl.root
	for {
		avl._propagate(node)
		t := int32(0)
		if node.left != nil {
			t = node.left.size
		}
		if t == k {
			return node.key
		} else if t < k {
			k -= t + 1
			node = node.right
		} else {
			node = node.left
		}
	}
}

func (avl *LazyAVLTree) _mergeWithRoot(l, root, r *avlNode) *avlNode {
	diff := int8(0)
	if l == nil {
		if r != nil {
			diff = -r.height
		}
	} else {
		if r == nil {
			diff = l.height
		} else {
			diff = l.height - r.height
		}
	}
	if diff > 1 {
		avl._propagate(l)
		l.right = avl._mergeWithRoot(l.right, root, r)
		avl._update(l)
		if l.left == nil {
			if l.right.height == 2 {
				return avl._balanceRight(l)
			}
		} else {
			if l.left.height-l.right.height == -2 {
				return avl._balanceRight(l)
			}
		}
		return l
	} else if diff < -1 {
		avl._propagate(r)
		r.left = avl._mergeWithRoot(l, root, r.left)
		avl._update(r)
		if r.right == nil {
			if r.left.height == 2 {
				return avl._balanceLeft(r)
			}
		} else {
			if r.left.height-r.right.height == 2 {
				return avl._balanceLeft(r)
			}
		}
		return r
	} else {
		root.left = l
		root.right = r
		avl._update(root)
		return root
	}
}

func (avl *LazyAVLTree) _mergeNode(l, r *avlNode) *avlNode {
	if l == nil {
		return r
	}
	if r == nil {
		return l
	}
	l, tmp := avl._popMax(l)
	return avl._mergeWithRoot(l, tmp, r)
}

func (avl *LazyAVLTree) _popMax(node *avlNode) (*avlNode, *avlNode) {
	avl._propagate(node)
	tmpPath = tmpPath[:0]
	path := tmpPath
	mx := node
	for node.right != nil {
		path = append(path, node)
		mx = node.right
		node = node.right
		avl._propagate(node)
	}
	path = append(path, node.left)
	len_ := len(path)
	for i := 0; i < len_-1; i++ {
		node = path[len(path)-1]
		path = path[:len(path)-1]
		if node == nil {
			path[len(path)-1].right = nil
			avl._update(path[len(path)-1])
			continue
		}
		b := avl._getBalance(node)
		if b == 2 {
			path[len(path)-1].right = avl._balanceLeft(node)
		} else if b == -2 {
			path[len(path)-1].right = avl._balanceRight(node)
		} else {
			path[len(path)-1].right = node
		}
		avl._update(path[len(path)-1])
	}
	if path[0] != nil {
		b := avl._getBalance(path[0])
		if b == 2 {
			path[0] = avl._balanceLeft(path[0])
		} else if b == -2 {
			path[0] = avl._balanceRight(path[0])
		}
	}
	mx.left = nil
	avl._update(mx)
	return path[0], mx
}

func (avl *LazyAVLTree) _splitNode(node *avlNode, k int32) (*avlNode, *avlNode) {
	if node == nil {
		return nil, nil
	}
	avl._propagate(node)
	tmp := k
	if node.left != nil {
		tmp -= node.left.size
	}
	if tmp == 0 {
		res1 := node.left
		res2 := avl._mergeWithRoot(nil, node, node.right)
		return res1, res2
	}
	if tmp < 0 {
		s, t := avl._splitNode(node.left, k)
		return s, avl._mergeWithRoot(t, node, node.right)
	}
	s, t := avl._splitNode(node.right, tmp-1)
	return avl._mergeWithRoot(node.left, node, s), t
}

func max8(x, y int8) int8 {
	if x > y {
		return x
	}
	return y
}

func max32(x, y int32) int32 {
	if x > y {
		return x
	}
	return y
}

func min32(x, y int32) int32 {
	if x < y {
		return x
	}
	return y
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func test() {
	arr := []int{1, 2, 3, 4, 5}
	tree := NewLazyAVLTree(int32(len(arr)), func(i int32) E { return E{arr[i], 1} })
	fmt.Println(tree.ToList())
	fmt.Println(tree.Pop(1))
	fmt.Println(tree.ToList())

	for i := 0; i < 1000; i++ {
		n := rand.Intn(1000) + 500
		arr = make([]int, n)
		for j := 0; j < n; j++ {
			arr[j] = rand.Intn(100) + 1
		}
		tree = NewLazyAVLTree(int32(len(arr)), func(i int32) E { return E{arr[i], 1} })

		for j := 0; j < 500; j++ {
			// set
			{
				index, value := rand.Intn(len(arr)), rand.Intn(100)+1
				arr[index] = value
				tree.Set(int32(index), E{value, 1})
			}

			// get
			{
				index := rand.Intn(len(arr))
				if arr[index] != int(tree.Get(int32(index)).sum) {
					panic("error get")
				}
			}

			// insert
			{
				index, value := rand.Intn(len(arr)+1), rand.Intn(100)+1
				arr = append(arr, 0)
				copy(arr[index+1:], arr[index:])
				arr[index] = value
				tree.Insert(int32(index), E{value, 1})

			}

			// pop
			{
				index := rand.Intn(len(arr))
				tmp1 := arr[index]
				tmp2 := tree.Pop(int32(index)).sum
				arr = append(arr[:index], arr[index+1:]...)
				if tmp1 != int(tmp2) {
					fmt.Println(tmp1, tmp2)
					panic("error pop")
				}
			}

			// apply
			{
				l, r, f := rand.Intn(len(arr)), rand.Intn(len(arr)), rand.Intn(100)+1
				for k := l; k < r; k++ {
					arr[k] += f
				}
				tree.Update(int32(l), int32(r), Id(f))
			}

			// apply all
			{
				f := rand.Intn(100) + 1
				for k := 0; k < len(arr); k++ {
					arr[k] += f
				}
				tree.UpdateAll(Id(f))
			}

			// prod
			{
				l, r := rand.Intn(len(arr)), rand.Intn(len(arr))
				if l > r {
					l, r = r, l
				}
				tmp1 := 0
				for i := l; i < r; i++ {
					tmp1 += arr[i]
				}
				tmp2 := tree.Query(int32(l), int32(r)).sum
				if tmp1 != int(tmp2) {
					fmt.Println(tmp1, tmp2)
					panic("error prod")
				}
			}

			// all prod
			{
				tmp1 := 0
				for i := 0; i < len(arr); i++ {
					tmp1 += arr[i]
				}
				tmp2 := tree.QueryAll().sum
				if tmp1 != int(tmp2) {
					fmt.Println(tmp1, tmp2)
					panic("error all prod")
				}
			}

			// tolist
			{

				tmp1 := make([]int, len(arr))
				copy(tmp1, arr)
				tmp2 := tree.ToList()
				if len(tmp1) != len(tmp2) {
					fmt.Println(tmp1, tmp2)
					panic("error tolist")
				}
				for i := 0; i < len(tmp1); i++ {
					if tmp1[i] != int(tmp2[i].sum) {
						fmt.Println(tmp1, tmp2)
						panic("error tolist")
					}
				}
			}

			// split + merge
			{

				l, r := rand.Intn(len(arr)), rand.Intn(len(arr))
				if l > r {
					l, r = r, l
				}
				lTree, rTree := tree.Split(int32(l))
				lTree.Merge(rTree)
				tree = lTree
			}
		}

	}

	fmt.Println("Passed")
}

func testTime() {
	n := int(2e5)
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = rand.Intn(100) + 1
	}

	tree := NewLazyAVLTree(int32(len(arr)), func(i int32) E { return E{arr[i], 1} })

	time1 := time.Now()
	for j := 0; j < int(2e5); j++ {
		tree.Set(int32(j), E{arr[j], 1})
		tree.Get(int32(j))
		tree.Insert(int32(j), E{arr[j], 1})
		tree.Pop(int32(j))
		tree.Update(int32(j), int32(n), Id(j))
		tree.UpdateAll(Id(j))
		tree.Query(int32(j), int32(n))
		tree.QueryAll()
		a, b := tree.Split(int32(j))
		a.Merge(b)
		tree = a
	}
	fmt.Println(time.Since(time1)) // 297.617458ms
}
